### ISSUE TITLE
Rust: 'Remove all unused imports' doesn't detect traits used in macros

### ISSUE THREAD

**Issue by bitfield** [zed-industries/zed#19269]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Repro:

```rust
use std::io::{self, Write, Result};

fn main() {
    writeln!(io::stdout(), "Hello, world!");
}
```

`rustc` correctly flags that the `Result` import is unused here, so let's go to it and use the code action "Remove all the unused imports". That leaves us with:

```rust
use std::io::{self}; // should really be just 'use std::io;', but never mind

fn main() {
    writeln!(io::stdout(), "Hello, world!");
}
```

Notice that the import of `Write` has been removed too, even though we actually need it: the call to `writeln!` doesn't work without it.

```
rustc: cannot write into `std::io::Stdout`
items from traits can only be used if the trait is in scope
```

If we're using `rust-analyzer` for this, then it's probably related to this bug: [Find all usages skips usages in derives](https://github.com/rust-lang/rust-analyzer/issues/15501).

Notably, when the trait usage is not hidden inside a macro, it works fine:

```rust
use std::io::{self, Write, Result};

fn main() {
    io::stdout().write_all("Hello, world!\n".as_bytes());
}
```

"Remove all unused imports" from this example correctly leaves `Write` in place, because it's required for `write_all`.

### Environment

Zed: v0.156.2 (Zed)
OS: macOS 15.0.1
Memory: 8 GiB
Architecture: aarch64

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>

**Comment by notpeter** [comment]
Yes, that action is from rust-analyzer, so it's like that issue upstream:
- https://github.com/rust-lang/rust-analyzer/issues/15501

Thanks for reporting.
